xhttp/xhttp.go

```diff
@@ -0,0 +1,57 @@
+package xhttp
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+type (
+	// Response is an extension of [http.Response] that contains parsed data
+	// instead of a response body.
+	Response[T any] struct {
+		*http.Response
+		Obj T
+	}
+	// ResponseErr is the error returned by [Do] if parsing the response body fails.
+	ResponseErr struct {
+		Err  error
+		Body []byte
+	}
+)
+
+// Do calls [Client.Do] and unmarshalls the HTTP response as a JSON of type [T].
+// The returned [Response] embeds the original [http.Response] with the addition
+// of the [Response.Obj] field that holds the parsed response.
+//
+// The original [http.Response.Body] will always be read and closed, the caller should ignore
+// this field and use [Response.Obj] to access the parsed response or use errors.As
+// to check details in the case of an error (eg. debugging malformed JSON).
+//
+// If the response is not valid JSON an error of type [ResponseErr] is returned.
+func Do[T any](c Client, req *http.Request) (*Response[T], error) {
```

Comment:
Thinking on leaving it just Do....we use JSON for 99% of HTTP interactions + there is no other reason that I can think off to have another layer on top of Do except this. But could also be DoJSON if someone feels like it would be better.

Reply:
we talked about this and I think the following would be useless but a nice addition haha:

```go
resp, _ := xhttp.Do[string](c, req)
if resp.Obj == "some string" {}
```

```go
resp, _ := xhttp.Do[[]byte](c, req)
if resp.Obj[0] == 10 {}
```
```
resp, _ := xhttp.Do[MyType](c, req)
if resp.Obj.Value == "some value" {}
```
etc

where `string` and `[]byte` would be special cases and all the rest handled as JSON.

but please ignore, JSON is used 100.00% of the time!

Reply:
Yeah that is a nice idea. But on the more pragmatic sense, lately I have been focusing more and more in a form of lightweight JSON RPC in my head. In the end HTTP/JSON won and everything uses it...and in general I like text protocols...and JSON is not great..but I suppose it could be worse ? :rofl: 

Of course sometimes you do need a binary protocol (like embeddings, we use JSON but it is painfully verbose for huge sets of numbers)...and for that I have no solution at all :laughing: (by not liking protobuf =P).